### PR TITLE
[train][release-test] Fix gptj finetuning example to be less sensitive to data ordering

### LINKED ISSUE #41391: Release test air_example_gptj_deepspeed_fine_tuning failed
Release test **air_example_gptj_deepspeed_fine_tuning** failed. See https://buildkite.com/ray-project/release/builds/2173#018c0eac-19be-4597-9712-7583322cf110 for more details.

Managed by OSS Test Policy

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

This PR fixes the `air_example_gptj_deepspeed_fine_tuning` release test to be less sensitive to data ordering. This test has failed in the past for a similar reason: Ray Data changes some implementation details, most likely causing the data ordering to change a little bit. Since this example uses fp16 weights to fit the GPTJ 6B model into a T4 16GB GPU, the learning dynamics are pretty brittle, and the gradients can blow up to NaN/inf pretty easily, just with a slightly different data ordering.

The previous fix (https://github.com/ray-project/ray/pull/40648) was to add a seeded `random_shuffle` to the dataset preprocessing pipeline that caused the fp16 training to run stably for an epoch. Since then, the test has started facing training instability again (and erroring out in the middle), so this PR introduces a longer term solution, which is to increase the tolerance of skipping batches when NaN/inf gradients appear. This is controlled by the `hysteresis` and `consecutive_hysteresis` deepspeed fp16 parameters:
* `hysteresis=4`: You must have 4 batches in a row give NaN/inf gradients before reducing the loss scale factor by 2
* `consecutive_hysteresis=True`: Any non-NaN/inf gradient batch will reset the counter back to 4.

## Related issue number

Closes https://github.com/ray-project/ray/issues/41391

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb [hunk 1]
```diff
@@ -486,7 +488,10 @@
     "    train_ds = train.get_dataset_shard(\"train\")\n",
     "    eval_ds = train.get_dataset_shard(\"validation\")\n",
     "\n",
-    "    train_ds_iterable = train_ds.iter_torch_batches(batch_size=batch_size)\n",
+    "    train_ds_iterable = train_ds.iter_torch_batches(\n",
+    "        batch_size=batch_size,\n",
+    "        local_shuffle_buffer_size=train.get_context().get_world_size() * batch_size,\n",
```
Comment (justinvyu) [inline doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:493]: local shuffle instead of global shuffle

File: doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb [hunk 2]
```diff
@@ -105,11 +105,11 @@
...
     "            \"evaluate\",\n",
-    "            # Latest combination of accelerate==0.19.0 and transformers==4.29.0\n",
-    "            # seems to have issues with DeepSpeed process group initialization,\n",
+    "            # The latest combination accelerate==0.25.0, transformers==4.36.0, deepspeed==0.12.4\n",
+    "            # has issues with DeepSpeed process group initialization,\n",
     "            # and will result in a batch_size validation problem.\n",
-    "            # TODO(jungong) : get rid of the pins once the issue is fixed.\n",
-    "            \"accelerate==0.16.0\",\n",
+    "            # TODO(ml-team): get rid of the pins once the issue is fixed.\n",
+    "            \"accelerate==0.18.0\",\n",
```
Comment (justinvyu) [inline doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:112]: the latest combo of libraries still doesn't work (the same batch size validation error as https://github.com/ray-project/ray/issues/35423)

File: doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb [hunk 3]
```diff
@@ -546,9 +551,37 @@
...
+    "tags": [
+     "remove-cell"
+    ]
+   },
+   "outputs": [],
+   "source": [
+    "# SMOKE TEST SETTINGS FOR CI\n",
+    "steps_per_epoch = 10\n",
+    "num_workers = 8\n",
+    "batch_size = 1"
```
Comment (justinvyu) [inline doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:584]: made the release test run for shorter/on fewer nodes

**Review by justinvyu — COMMENTED** [review]
notes on some other changes

**Comment by justinvyu** [comment]
Passing release test here: https://buildkite.com/ray-project/release/builds/4308#018c6aa6-6f5a-4a3b-84c2-5bf94009a15d

File: doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb [hunk 4]
```diff
@@ -413,6 +415,8 @@
     "        \"fp16\": {\n",
     "            \"enabled\": \"auto\",\n",
     "            \"initial_scale_power\": 8,\n",
+    "            \"hysteresis\": 4,\n",
```
Comment (woshiyyya) [inline doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:418]: What's the behavior of `hysteresis` and `consecutive_hysteresis`?

**Review by woshiyyya — APPROVED** [review]

File: doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb (hunk 4 above)
Comment (justinvyu) [reply doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:418]: Updated the PR description:

> * `hysteresis=4`: You must have 4 batches in a row give NaN/inf gradients before reducing the loss scale factor by 2
> * `consecutive_hysteresis=True`: Any non-NaN/inf gradient batch will reset the counter back to 4.

Comment (woshiyyya) [reply doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:418]: I see. So does that means, if we don't set `hysteresis`, deepspeed will try to scale loss factor by 2 whenever it encounters a batch with nan/inf gradient. 
After we set these params, it will skip the batch with Nan/inf gradients 4 times before rescale the loss?

Comment (justinvyu) [reply doc/source/train/examples/deepspeed/gptj_deepspeed_fine_tuning.ipynb:418]: Exactly, we require 4 consecutive NaN batches in order to decrease the loss scale by 1/2. This basically skips more batches in the long run, but results in training that doesn't immediately error out due to getting to the minimum loss scale. Definitely not ideal though.
